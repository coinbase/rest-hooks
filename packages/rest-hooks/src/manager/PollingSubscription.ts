import { Schema } from '@rest-hooks/normalizr';
import { actionTypes, Dispatch } from '@rest-hooks/core';

import { Subscription, SubscriptionInit } from './SubscriptionManager';
import DefaultConnectionListener from './DefaultConnectionListener';
import ConnectionListener from './ConnectionListener';

const { FETCH_TYPE } = actionTypes;

/**
 * PollingSubscription keeps a given resource updated by
 * dispatching a fetch at a rate equal to the minimum update
 * interval requested.
 */
export default class PollingSubscription implements Subscription {
  protected declare readonly schema: Schema;
  protected declare readonly fetch: () => Promise<any>;
  protected declare readonly key: string;
  protected declare frequency: number;
  protected frequencyHistogram: Map<number, number> = new Map();
  protected declare dispatch: Dispatch<any>;
  protected declare intervalId?: NodeJS.Timeout;
  protected declare lastIntervalId?: NodeJS.Timeout;
  private declare connectionListener: ConnectionListener;

  constructor(
    { key, schema, fetch, frequency }: SubscriptionInit,
    dispatch: Dispatch<any>,
    connectionListener?: ConnectionListener,
  ) {
    if (frequency === undefined)
      throw new Error('frequency needed for polling subscription');
    this.schema = schema;
    this.fetch = fetch;
    this.frequency = frequency;
    this.key = key;
    this.frequencyHistogram.set(this.frequency, 1);
    this.dispatch = dispatch;
    this.connectionListener =
      connectionListener || new DefaultConnectionListener();

    // Kickstart running since this is initialized after the online notif is sent
    if (this.connectionListener.isOnline()) {
      this.onlineListener();
    } else {
      this.offlineListener();
    }
  }

  /** Subscribe to a frequency */
  add(frequency?: number) {
    if (frequency === undefined) return;
    if (this.frequencyHistogram.has(frequency)) {
      this.frequencyHistogram.set(
        frequency,
        (this.frequencyHistogram.get(frequency) as number) + 1,
      );
    } else {
      this.frequencyHistogram.set(frequency, 1);

      // new min so restart service
      if (frequency < this.frequency) {
        this.frequency = frequency;
        this.run();
      }
    }
  }

  /** Unsubscribe from a frequency */
  remove(frequency?: number) {
    if (frequency === undefined) return false;
    if (this.frequencyHistogram.has(frequency)) {
      this.frequencyHistogram.set(
        frequency,
        (this.frequencyHistogram.get(frequency) as number) - 1,
      );
      if ((this.frequencyHistogram.get(frequency) as number) < 1) {
        this.frequencyHistogram.delete(frequency);

        // nothing subscribed to this anymore...it is invalid
        if (this.frequencyHistogram.size === 0) {
          this.cleanup();
          return true;
        }

        // this was the min, so find the next size
        if (frequency <= this.frequency) {
          this.frequency = Math.min(...this.frequencyHistogram.keys());
          this.run();
        }
      }
    } /* istanbul ignore next */ else if (
      process.env.NODE_ENV !== 'production'
    ) {
      console.error(
        `Mismatched remove: ${frequency} is not subscribed for ${this.key}`,
      );
    }
    return false;
  }

  /** Cleanup means clearing out background interval. */
  cleanup() {
    if (this.intervalId) {
      clearInterval(this.intervalId);
      this.intervalId = undefined;
    }
    if (this.lastIntervalId) {
      clearInterval(this.lastIntervalId);
      this.lastIntervalId = undefined;
    }
    this.connectionListener.removeOnlineListener(this.onlineListener);
    this.connectionListener.removeOfflineListener(this.offlineListener);
  }

  /** Trigger request for latest resource */
  protected update() {
    this.dispatch({
      type: FETCH_TYPE,
      payload: this.fetch,
      meta: {
        schema: this.schema,
        key: this.key,
        type: 'read',
        throttle: true,
        options: {
          dataExpiryLength: this.frequency / 2,
          errorExpiryLength: this.frequency / 10,
        },
        resolve: () => {},
        reject: () => {},
      },
    });
  }

  /** What happens when browser goes offline */
  protected offlineListener = () => {
    // this clears existing listeners, so no need to clear offline listener
    this.cleanup();
    this.connectionListener.addOnlineListener(this.onlineListener);
  };

  /** What happens when browser comes online */
  protected onlineListener = () => {
    this.connectionListener.removeOnlineListener(this.onlineListener);
    this.update();
    this.run();
    this.connectionListener.addOfflineListener(this.offlineListener);
  };

  /** Run polling process with current frequency
   *
   * Will clean up old poll interval on next run
   */
  protected run() {
    this.lastIntervalId = this.intervalId;
    this.intervalId = setInterval(() => {
      // since we don't know how long into the last poll it was before resetting
      // we wait til the next fetch to clear old intervals
      if (this.lastIntervalId) {
        clearInterval(this.lastIntervalId);
        this.lastIntervalId = undefined;
      }
      this.update();
    }, this.frequency);
  }
}
