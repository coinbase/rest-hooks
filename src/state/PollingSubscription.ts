import { Schema } from '../resource';
import { Subscription, SubscriptionInit } from './SubscriptionManager';

function min(iter: IterableIterator<number>) {
  let ret = Infinity;
  for (const n of iter) {
    if (n < ret) ret = n;
  }
  return ret;
}

/**
 * PollingSubscription keeps a given resource updated by
 * dispatching a fetch at a rate equal to the minimum update
 * interval requested.
 */
export default class PollingSubscription implements Subscription {
  protected readonly schema: Schema;
  protected readonly fetch: () => Promise<any>;
  protected readonly url: string;
  protected frequency: number;
  protected frequencyHistogram: Map<number, number> = new Map();
  protected dispatch: React.Dispatch<any>;
  protected intervalId?: NodeJS.Timeout;
  protected lastIntervalId?: NodeJS.Timeout;

  constructor(
    { url, schema, fetch, frequency }: SubscriptionInit,
    dispatch: React.Dispatch<any>,
  ) {
    if (frequency === undefined)
      throw new Error('frequency needed for polling subscription');
    this.schema = schema;
    this.fetch = fetch;
    this.frequency = frequency;
    this.url = url;
    this.frequencyHistogram.set(this.frequency, 1);
    this.dispatch = dispatch;
    this.run();
  }

  /** Subscribe to a frequency */
  add(frequency?: number) {
    if (frequency === undefined) return;
    if (this.frequencyHistogram.has(frequency)) {
      this.frequencyHistogram.set(
        frequency,
        (this.frequencyHistogram.get(frequency) as number) + 1,
      );
    } else {
      this.frequencyHistogram.set(frequency, 1);

      // new min so restart service
      if (frequency < this.frequency) {
        this.frequency = frequency;
        this.run();
      }
    }
  }

  /** Unsubscribe from a frequency */
  remove(frequency?: number) {
    if (frequency === undefined) return false;
    if (this.frequencyHistogram.has(frequency)) {
      this.frequencyHistogram.set(
        frequency,
        (this.frequencyHistogram.get(frequency) as number) - 1,
      );
      if ((this.frequencyHistogram.get(frequency) as number) < 1) {
        this.frequencyHistogram.delete(frequency);

        // this was the min, so find the next size
        if (frequency <= this.frequency) {
          if (this.frequencyHistogram.size === 0) return true;
          this.frequency = min(this.frequencyHistogram.keys());
          this.run();
        }
      }
    } else if (process.env.NODE_ENV !== 'production') {
      console.error(
        `Mismatched remove: ${this.frequency} is not subscribed for ${
          this.url
        }`,
      );
    }
    return false;
  }

  /** Cleanup means clearing out background interval. */
  cleanup() {
    if (this.intervalId) clearInterval(this.intervalId);
    if (this.lastIntervalId) clearInterval(this.lastIntervalId);
  }

  /** Trigger request for latest resource */
  protected update() {
    this.dispatch({
      type: 'fetch',
      payload: this.fetch,
      meta: {
        schema: this.schema,
        url: this.url,
        responseType: 'receive',
        throttle: true,
        options: {
          dataExpiryLength: this.frequency / 2,
          errorExpiryLength: this.frequency / 10,
        },
      },
    });
  }

  /** Run polling process with current frequency
   *
   * Will clean up old poll interval on next run
   */
  protected run() {
    this.lastIntervalId = this.intervalId;
    this.intervalId = setInterval(() => {
      // since we don't know how long into the last poll it was before resetting
      // we wait til the next fetch to clear old intervals
      if (this.lastIntervalId) {
        clearInterval(this.lastIntervalId);
        this.lastIntervalId = undefined;
      }
      this.update();
    }, this.frequency);
  }
}
