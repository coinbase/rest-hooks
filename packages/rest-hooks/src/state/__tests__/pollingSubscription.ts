import { PollingArticleResource } from '__tests__/common';

import PollingSubscription from '../PollingSubscription';
import { ConnectionListener } from '../ConnectionListener';

class MockConnectionListener implements ConnectionListener {
  declare online: boolean;
  declare onlineHandlers: (() => void)[];
  declare offlineHandlers: (() => void)[];

  constructor(online: boolean) {
    this.online = online;
    this.onlineHandlers = [];
    this.offlineHandlers = [];
  }

  isOnline() {
    return this.online;
  }

  addOnlineListener(handler: () => void) {
    this.onlineHandlers.push(handler);
  }

  removeOnlineListener(handler: () => void) {
    this.onlineHandlers = this.onlineHandlers.filter(h => h !== handler);
  }

  addOfflineListener(handler: () => void) {
    this.offlineHandlers.push(handler);
  }

  removeOfflineListener(handler: () => void) {
    this.offlineHandlers = this.offlineHandlers.filter(h => h !== handler);
  }

  trigger(event: 'offline' | 'online') {
    switch (event) {
      case 'offline':
        this.online = false;
        this.offlineHandlers.forEach(t => t());
        break;
      case 'online':
        this.online = true;
        this.onlineHandlers.forEach(t => t());
        break;
    }
  }

  reset() {
    this.offlineHandlers = [];
    this.onlineHandlers = [];
  }
}

function onError(e: any) {
  e.preventDefault();
}
beforeAll(() => {
  window.addEventListener('error', onError);
});
afterAll(() => {
  window.removeEventListener('error', onError);
});

describe('PollingSubscription', () => {
  const dispatch = jest.fn();
  const a = () => Promise.resolve({ id: 5, title: 'hi' });
  const fetch = jest.fn(a);
  jest.useFakeTimers();

  const sub = new PollingSubscription(
    {
      url: 'test.com',
      schema: PollingArticleResource.getEntitySchema(),
      fetch,
      frequency: 5000,
    },
    dispatch,
  );

  afterAll(() => {
    sub.cleanup();
  });

  it('should throw on undefined frequency in construction', () => {
    expect(
      () =>
        new PollingSubscription(
          {
            url: 'test.com',
            schema: PollingArticleResource.getEntitySchema(),
            fetch,
          },
          dispatch,
        ),
    ).toThrow();
  });

  it('should call immediately', () => {
    expect(dispatch.mock.calls.length).toBe(1);
  });

  it('should call after period', () => {
    dispatch.mockReset();
    jest.advanceTimersByTime(5000);
    expect(dispatch.mock.calls.length).toBe(1);
    expect(dispatch.mock.calls[0]).toMatchSnapshot();
    jest.advanceTimersByTime(5000);
    expect(dispatch.mock.calls.length).toBe(2);
    expect(dispatch.mock.calls[1]).toMatchSnapshot();
  });

  describe('add()', () => {
    it('should take smaller frequency when another is added', () => {
      sub.add(1000);
      jest.advanceTimersByTime(1000 * 4);
      expect(dispatch.mock.calls.length).toBe(2 + 4);
    });

    it('should not change frequency if same is added', () => {
      dispatch.mockClear();
      sub.add(1000);
      jest.advanceTimersByTime(1000 * 13);
      expect(dispatch.mock.calls.length).toBe(13);
    });

    it('should not change frequency if larger is added', () => {
      dispatch.mockClear();
      sub.add(7000);
      sub.add(8000);
      jest.advanceTimersByTime(1000 * 13);
      expect(dispatch.mock.calls.length).toBe(13);
    });

    it('should do nothing if frequency is undefined', () => {
      dispatch.mockClear();
      sub.add(undefined);
      jest.advanceTimersByTime(1000 * 13);
      expect(dispatch.mock.calls.length).toBe(13);
    });
  });

  describe('remove()', () => {
    it('should not change frequency if only partially removed', () => {
      dispatch.mockClear();
      sub.remove(1000);
      jest.advanceTimersByTime(1000 * 13);
      expect(dispatch.mock.calls.length).toBe(13);
    });

    it('should not change frequency if only larger removed', () => {
      dispatch.mockClear();
      sub.remove(7000);
      jest.advanceTimersByTime(1000 * 13);
      expect(dispatch.mock.calls.length).toBe(13);
    });

    it('should go back to fastest if smallest frequency is removed completely', () => {
      sub.remove(1000);
      jest.advanceTimersByTime(1000);
      dispatch.mockClear();
      jest.advanceTimersByTime(5000 * 13);
      expect(dispatch.mock.calls.length).toBe(13);
    });

    it('should do nothing if frequency is not registered', () => {
      const oldError = console.error;
      const spy = (console.error = jest.fn());

      sub.remove(1000);
      dispatch.mockClear();
      jest.advanceTimersByTime(5000 * 13);
      expect(dispatch.mock.calls.length).toBe(13);

      expect(spy.mock.calls[0]).toMatchInlineSnapshot(`
        Array [
          "Mismatched remove: 1000 is not subscribed for test.com",
        ]
      `);
      console.error = oldError;
    });

    it('should do nothing if frequency is undefined', () => {
      sub.remove(undefined);
      dispatch.mockClear();
      jest.advanceTimersByTime(5000 * 13);
      expect(dispatch.mock.calls.length).toBe(13);
    });

    it('should return false until completely empty, then return true', () => {
      const sub = new PollingSubscription(
        {
          url: 'test.com',
          schema: PollingArticleResource.getEntitySchema(),
          fetch,
          frequency: 5000,
        },
        dispatch,
      );
      sub.add(5000);
      sub.add(7000);
      expect(sub.remove(5000)).toBe(false);
      expect(sub.remove(5000)).toBe(false);
      expect(sub.remove(7000)).toBe(true);
    });
  });

  describe('cleanup()', () => {
    it('should stop all timers', () => {
      dispatch.mockClear();
      sub.cleanup();
      jest.advanceTimersByTime(5000 * 13);
      expect(dispatch.mock.calls.length).toBe(0);
    });
    it('should be idempotent', () => {
      sub.cleanup();
    });
  });

  describe('offline support', () => {
    jest.useFakeTimers();

    function createMocks(listener: ConnectionListener) {
      const dispatch = jest.fn();
      const a = () => Promise.resolve({ id: 5, title: 'hi' });
      const fetch = jest.fn(a);

      const pollingSubscription = new PollingSubscription(
        {
          url: 'test.com',
          schema: PollingArticleResource.getEntitySchema(),
          fetch,
          frequency: 5000,
        },
        dispatch,
        listener,
      );
      return { dispatch, fetch, pollingSubscription };
    }

    it('should not dispatch when offline', () => {
      const listener = new MockConnectionListener(false);
      const { dispatch } = createMocks(listener);
      jest.advanceTimersByTime(50000);
      expect(dispatch.mock.calls.length).toBe(0);
      expect(listener.offlineHandlers.length).toBe(0);
      expect(listener.onlineHandlers.length).toBe(1);
    });

    it('should immediately start fetching when online', () => {
      const listener = new MockConnectionListener(false);
      const { dispatch } = createMocks(listener);
      expect(dispatch.mock.calls.length).toBe(0);

      listener.trigger('online');
      expect(dispatch.mock.calls.length).toBe(1);
      jest.advanceTimersByTime(5000);
      expect(dispatch.mock.calls.length).toBe(2);
      expect(listener.offlineHandlers.length).toBe(1);
      expect(listener.onlineHandlers.length).toBe(0);
    });

    it('should stop dispatching when offline again', () => {
      const listener = new MockConnectionListener(true);
      const { dispatch } = createMocks(listener);
      expect(dispatch.mock.calls.length).toBe(1);

      listener.trigger('offline');
      jest.advanceTimersByTime(50000);
      expect(dispatch.mock.calls.length).toBe(1);
      expect(listener.offlineHandlers.length).toBe(0);
      expect(listener.onlineHandlers.length).toBe(1);
    });
  });
});
