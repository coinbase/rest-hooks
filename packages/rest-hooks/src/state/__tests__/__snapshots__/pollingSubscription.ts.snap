// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollingSubscription should call after period 1`] = `
Array [
  Object {
    "meta": Object {
      "options": Object {
        "dataExpiryLength": 2500,
        "errorExpiryLength": 500,
      },
      "reject": [Function],
      "resolve": [Function],
      "responseType": "rest-hooks/receive",
      "schema": [Function],
      "throttle": true,
      "url": "test.com",
    },
    "payload": [MockFunction],
    "type": "rest-hooks/fetch",
  },
]
`;

exports[`PollingSubscription should call after period 2`] = `
Array [
  Object {
    "meta": Object {
      "options": Object {
        "dataExpiryLength": 2500,
        "errorExpiryLength": 500,
      },
      "reject": [Function],
      "resolve": [Function],
      "responseType": "rest-hooks/receive",
      "schema": [Function],
      "throttle": true,
      "url": "test.com",
    },
    "payload": [MockFunction],
    "type": "rest-hooks/fetch",
  },
]
`;
